src/v/archival/ntp_archiver_service.cc

```diff
@@ -2608,6 +2678,112 @@ ss::future<> ntp_archiver::apply_spillover() {
     }
 }
 
+flush_result ntp_archiver::flush() {
+    // Return early if we are not the leader, or if we are a read replica.
+    if (!_parent.is_leader() || _parent.is_read_replica_mode_enabled()) {
+        vlog(
+          _rtclog.debug,
+          "Flush request was not accepted, as partition is not the leader or "
+          "is a read replica");
+        return flush_result{
+          .response = flush_response::rejected, .offset = std::nullopt};
+    }
+
+    _flush_uploads_offset = model::prev_offset(
+      max_uploadable_offset_exclusive());
+    vlog(
+      _rtclog.debug,
+      "Accepted flush, flush offset is {}",
+      _flush_uploads_offset.value());
+    return flush_result{
+      .response = flush_response::accepted,
+      .offset = _flush_uploads_offset.value()};
+}
+
+ss::future<wait_result> ntp_archiver::wait(model::offset o) {
+    if (_parent.is_read_replica_mode_enabled()) {
+        vlog(
+          _rtclog.debug,
+          "Cannot wait on a flush in ntp_archiver, as this partition is a read "
+          "replica");
+        co_return wait_result::failed;
+    }
+
+    // We have uploaded or flush'ed() up to this point.
+    if (uploaded_and_clean_past_offset(o)) {
+        vlog(
+          _rtclog.debug,
+          "Already uploaded past passed offset {} before waiting on a flush "
+          "in ntp_archiver",
+          o);
+        co_return wait_result::complete;
+    }
+
+    // If this partition has lost leadership, we indicate so.
+    if (lost_leadership()) {
+        vlog(
+          _rtclog.debug,
+          "Before waiting on a flush in ntp_archiver, leadership was already "
+          "lost");
+        co_return wait_result::lost_leadership;
+    }
+
+    // Currently we tie wait() to flush() state. If there is no flush in
+    // progress, we return.
+    if (!flush_in_progress()) {
+        vlog(
+          _rtclog.debug,
+          "Cannot wait on a flush in ntp_archiver, as no flush in progress");
+        co_return wait_result::not_in_progress;
+    }
+
+    // If o is outside bounds of _flush_uploads_offset, we indicate so.
+    if (o > _flush_uploads_offset.value()) {
+        vlog(
+          _rtclog.debug,
+          "Passed offset {} is outside bounds of current flush offset {}",
+          o,
+          _flush_uploads_offset.value());
+        co_return wait_result::not_in_progress;
+    }
+
+    // Save the current_term before suspending on condition variable.
+    auto current_term = _start_term;
+
+    // Otherwise, we will wait on the condition variable to signal.
+    try {
+        vlog(_rtclog.debug, "Waiting on flush to complete in ntp_archiver");
+        co_await _flush_cond.wait();
```

Comment:
another option is to restructure part of this method as a loop and restart the loop when the condition is met so that checks are not duplicated below.

Reply:
Overhauled this function to use a `while()` loop, resulting in much nicer logic and function size.